apiserver/facades/agent/caasagent/caasagent.go

```diff
@@ -4,14 +4,173 @@
+	APIHostPortsForAgents(controller.Config) ([]network.SpaceHostPorts, error)
+	CompletedMigrationForModel(string) (state.ModelMigration, error)
+}
+
+// NewFacadeV2 creates a new caas admission facade v2.
+func NewFacadeV2(
+	modelUUID coremodel.UUID,
+	registry facade.WatcherRegistry,
+	controllerConfigService ControllerConfigService,
+	modelConfigService ModelConfigService,
+	externalControllerService ExternalControllerService,
+	controllerConfigState ControllerConfigState,
+	cloudSpecGetter StubService,
+	modelService ModelService,
+) *FacadeV2 {
+	return &FacadeV2{
+		modelUUID:          modelUUID,
+		registry:           registry,
+		cloudSpecGetter:    cloudSpecGetter,
+		modelService:       modelService,
+		ModelConfigWatcher: commonmodel.NewModelConfigWatcher(modelConfigService, registry),
+		ControllerConfigAPI: common.NewControllerConfigAPI(
+			controllerConfigState,
+			controllerConfigService,
+			externalControllerService,
+		),
+	}
+}
+
+// FacadeV2 is the V2 facade of the caas agent.
 type FacadeV2 struct {
-	cloudspec.CloudSpecer
-	*model.ModelConfigWatcher
+	*commonmodel.ModelConfigWatcher
 	*common.ControllerConfigAPI
+
+	registry        facade.WatcherRegistry
+	modelUUID       coremodel.UUID
+	cloudSpecGetter StubService
+	modelService    ModelService
+}
+
+// CloudSpec returns the cloud spec used by the specified models.
+func (f *FacadeV2) CloudSpec(ctx context.Context, args params.Entities) (params.CloudSpecResults, error) {
+	results := params.CloudSpecResults{
+		Results: make([]params.CloudSpecResult, len(args.Entities)),
+	}
+	for i, arg := range args.Entities {
+		tag, err := names.ParseModelTag(arg.Tag)
+		if err != nil {
+			results.Results[i].Error = apiservererrors.ServerError(err)
+			continue
+		}
+		if tag.Id() != f.modelUUID.String() {
+			results.Results[i].Error = apiservererrors.ServerError(apiservererrors.ErrPerm)
+			continue
+		}
+		spec, err := f.cloudSpecGetter.CloudSpec(ctx)
+		if err != nil {
+			results.Results[i].Error = apiservererrors.ServerError(err)
+			continue
+		}
+		result := params.CloudSpecResult{
+			Error: apiservererrors.ServerError(err),
+		}
+		if err == nil {
+			result.Result = common.CloudSpecToParams(spec)
+		}
+		results.Results[i] = result
+	}
+	return results, nil
+}
+
+// WatchCloudSpecsChanges returns a watcher for cloud spec changes.
+func (f *FacadeV2) WatchCloudSpecsChanges(ctx context.Context, args params.Entities) (params.NotifyWatchResults, error) {
+	results := params.NotifyWatchResults{
+		Results: make([]params.NotifyWatchResult, len(args.Entities)),
+	}
+	for i, arg := range args.Entities {
+		tag, err := names.ParseModelTag(arg.Tag)
+		if err != nil {
+			results.Results[i].Error = apiservererrors.ServerError(err)
+			continue
+		}
+		if tag.Id() != f.modelUUID.String() {
+			results.Results[i].Error = apiservererrors.ServerError(apiservererrors.ErrPerm)
+			continue
+		}
+		w, err := f.modelService.WatchModelCloudCredential(ctx, f.modelUUID)
```

Comment:
There can only be one watcher resulting from this call. i.e you can either pass model id "123" once or 6 thousand times. We still only need the one watcher.

Can we just request this once and re-use? Ideally the for loop just becomes a sanity check that you have only request 1 model in the args etc.

Reply:
Added a wrapper and a check for dupes